File: Algorytmy rozwiązywania gier/players.py
from colors import Color
import copy
from estimators import Estimator
from datetime import datetime


class Player:
    def __init__(self, colour, name=None):
        from checkers import Checkers
        if colour not in [Checkers.WHITE, Checkers.BLACK]:
            raise Exception(f'Invalid colour. Possible options are: {Checkers.WHITE} and {Checkers.BLACK}.')
        self.colour = colour
        self.name = self.__class__.__name__ + self.colour if name is None else name

    def move(self, possible_moves, board) -> int:
        return 0

    def capture(self, possible_moves, board) -> int:
        return 0


class Human(Player):
    def __init__(self, colour, name=None):
        Player.__init__(self, colour, name)

    def move(self, possible_moves, board) -> int:
        print(f'{Color.color(153, 255, 51)}{self.name} moves:{Color.END}')
        from checkers import Checkers as C
        for i, m in enumerate(possible_moves):
            start = C.alias_from_coordinates(m[0])
            end = C.alias_from_coordinates(m[1])
            print(f'{i}. From {start} to {end}')
        if not possible_moves:
            print('0. skip')
        is_correct = False
        pos = 0
        while not is_correct:
            try:
                pos = int(input('Move: '))
                if not possible_moves:
                    if pos == 0:
                        is_correct = True
                        pos = -1
                elif 0 <= pos < len(possible_moves):
                    is_correct = True
                else:
                    print(f'Invalid input. Type number between 0 and {len(possible_moves) - 1}')
            except ValueError:
                print(f'Invalid input. Type number between 0 and {len(possible_moves) - 1}')
        return pos

    def capture(self, possible_captures, board) -> int:
        from checkers import Checkers as C
        print(f'{Color.color(255, 128, 0)}{self.name} captures:{Color.END}')
        for i, m in enumerate(possible_captures):
            start = C.alias_from_coordinates(m[0])
            path = [C.alias_from_coordinates(coor) for coor in m[1:]]
            print(f'{i}. From {start} over {path}')
        is_correct = False
        pos = 0
        while not is_correct:
            try:
                pos = int(input('Move: '))
                if 0 <= pos < len(possible_captures):
                    is_correct = True
                else:
                    print(f'Invalid input. Type number between 0 and {len(possible_captures) - 1}')
            except ValueError:
                print(f'Invalid input. Type number between 0 and {len(possible_captures) - 1}')
        return pos


class DummyBot(Player):

    NAMES = ['Niedorzeczny Bóbr',
             'Zgryźliwa Żmija',
             'Szybki Sebastian',
             'Pijany Olek',
             'Potężna Grochówka',
             'Smutny Marek',
             'Wesoły Romek']

    def __init__(self, colour, name=None):
        import random
        n = random.choice(self.NAMES) if name is None else name
        Player.__init__(self, colour, n)

    def move(self, possible_moves, board) -> int:
        if not possible_moves:
            return -1
        if len(possible_moves) == 1:
            return 0
        import random
        return random.randint(0, len(possible_moves) - 1)

    def capture(self, possible_moves, board) -> int:
        if not possible_moves:
            return -1
        if len(possible_moves) == 1:
            return 0
        import random
        return random.randint(0, len(possible_moves) - 1)


class MinMaxBot(Player):
    NAMES = ['Minimalistyczny Bóbr',
             'Optymalizacyjna Żmija',
             'Rekurencyjny Sebastian',
             'Maksujący Olek',
             'Zbijająca Grochówka',
             'Powolny Marek',
             'Drzewny Romek']

    MAX_VALUE = int(1e9)
    MIN_VALUE = int(-1e9)

    def __init__(self, colour, estimator: Estimator, search_depth=3, name=None, time_constraint=int(1e10)):
        import random
        from checkers import Checkers
        n = random.choice(self.NAMES) if name is None else name
        Player.__init__(self, colour, n)
        self.colour = colour
        self.opposite_colour = Checkers.WHITE if self.colour == Checkers.BLACK else Checkers.BLACK
        self.search_depth = search_depth
        self.estimator = estimator
        self.time_constraint = time_constraint

    def move(self, possible_moves, board) -> int:
        start = datetime.now()
        pos = self.get_best_move(possible_moves, board)
        end = datetime.now()
        print(f'Computed in {(end - start).seconds}s')
        return pos

    def capture(self, possible_moves, board) -> int:
        start = datetime.now()
        pos = self.get_best_capture(possible_moves, board)
        end = datetime.now()
        print(f'Computed in {(end - start).seconds}s')
        return pos

    def get_best_capture(self, possible_captures, board) -> int:
        if not possible_captures:
            return -1
        if len(possible_captures) == 1:
            return 0
        best_captures = []
        best_result = 0
        from checkers import Checkers
        for i, capture in enumerate(possible_captures):
            new_board = copy.deepcopy(board)
            new_board = Checkers.API.execute_capture(new_board, self, capture)
            start = datetime.now()
            estimation = self.min(new_board, self.search_depth)
            end = datetime.now()
            if not best_captures or estimation == best_result:
                best_captures.append(i)
            elif estimation > best_result:
                best_captures = [i]
                best_result = estimation
            if (end - start).seconds > self.time_constraint:
                break
        import random
        if not best_captures:
            return random.choice([i for i in range(len(possible_captures))])
        return random.choice(best_captures)

    def get_best_move(self, possible_moves, board) -> int:
        if not possible_moves:
            return -1
        if len(possible_moves) == 1:
            return 0
        best_moves = []
        best_result = 0
        from checkers import Checkers
        for i, move in enumerate(possible_moves):
            new_board = copy.deepcopy(board)
            new_board = Checkers.API.execute_move(new_board, self, move)
            start = datetime.now()
            estimation = self.min(new_board, self.search_depth)
            if not best_moves or estimation == best_result:
                best_moves.append(i)
            elif estimation > best_result:
                best_moves = [i]
                best_result = estimation
            end = datetime.now()
            if (end - start).seconds > self.time_constraint:
                break
        import random
        if not best_moves:
            return random.choice([i for i in range(len(possible_moves))])
        return random.choice(best_moves)

    def min(self, board, depth) -> int:
        current_estimation = self.estimator(board, Player(self.opposite_colour))
        if depth <= 0:
            return current_estimation
        from checkers import Checkers
        if Checkers.API.has_game_ended(board):
            return current_estimation
        possible_captures = Checkers.API.all_captures(board, Player(self.opposite_colour))
        possible_moves = Checkers.API.all_moves(board, Player(self.opposite_colour))
        if not possible_moves and not possible_captures:
            return self.MAX_VALUE
        worst_result = self.MAX_VALUE
        if possible_captures:
            for capture in possible_captures:
                new_board = copy.deepcopy(board)
                new_board = Checkers.API.execute_capture(new_board, Player(self.opposite_colour), capture)
                estimation = self.max(new_board, depth - 1)
                if worst_result > estimation:
                    worst_result = estimation
            return worst_result
        for move in possible_moves:
            new_board = copy.deepcopy(board)
            new_board = Checkers.API.execute_move(new_board, Player(self.opposite_colour), move)
            estimation = self.max(new_board, depth - 1)
            if worst_result > estimation:
                worst_result = estimation
        return worst_result

    def max(self, board, depth) -> int:
        current_estimation = self.estimator(board, self)
        if depth <= 0:
            return current_estimation
        from checkers import Checkers
        if Checkers.API.has_game_ended(board):
            return current_estimation
        possible_captures = Checkers.API.all_captures(board, self)
        possible_moves = Checkers.API.all_moves(board, self)
        if not possible_moves and not possible_captures:
            return self.MIN_VALUE
        best_result = self.MIN_VALUE
        if possible_captures:
            for capture in possible_captures:
                new_board = copy.deepcopy(board)
                new_board = Checkers.API.execute_capture(new_board, self, capture)
                estimation = self.min(new_board, depth - 1)
                if best_result < estimation:
                    best_result = estimation
            return best_result
        for move in possible_moves:
            new_board = copy.deepcopy(board)
            new_board = Checkers.API.execute_move(new_board, self, move)
            estimation = self.min(new_board, depth - 1)
            if best_result < estimation:
                best_result = estimation
        return best_result

File: Algorytmy rozwiazywania gier kotlin/src/players.kt
import java.lang.Integer.min
import java.lang.System.currentTimeMillis
import kotlin.math.max
import kotlin.reflect.jvm.internal.impl.util.Check

fun <T> getCorrectInput(list: List<T>): Int {
    var isCorrect = false
    var pos = 0
    while(!isCorrect) {
        try {
            pos = readLine()!!.toInt()
            if(pos >= 0 && pos < list.size) {
                isCorrect = true
            } else {
                println("Invalid input. Type number between 0 and ${list.size - 1}")
            }
        } catch (e: Exception) {
            println("Invalid input. Type number between 0 and ${list.size - 1}")
        }
    }
    return pos
}


open class Player(open val name: String = "${Player::class.simpleName}") {

    lateinit var colour: CheckersColour

    open fun move(possibleMoves: List<Move>, board: Board): Int {
        return 0
    }

    open fun capture(possibleCaptures: List<List<Jump>>, board: Board): Int {
        return 0
    }

    override fun toString(): String {
        return name
    }
}

class Human(name: String): Player(name) {
    override fun move(possibleMoves: List<Move>, board: Board): Int {
        println("${Colour.colour(153, 255, 51)}${name} moves:${Colour.END}")
        for((i, move) in possibleMoves.withIndex()){
            val start = CheckersController.aliasFromCoordinates(move.xStart, move.yStart)
            val end = CheckersController.aliasFromCoordinates(move.xEnd, move.yEnd)
            println("${i}. From $start to $end")
        }
        return getCorrectInput(possibleMoves)
    }

    override fun capture(possibleCaptures: List<List<Jump>>, board: Board): Int {
        println("${Colour.colour(153, 255, 51)}${name} moves:${Colour.END}")
        for((i, jumps) in possibleCaptures.withIndex()) {
            val startJump = jumps[0]
            val start = CheckersController.aliasFromCoordinates(startJump.xStart, startJump.yStart)
            var path = ""
            for(jump in jumps) {
                path += " -> "
                path += CheckersController.aliasFromCoordinates(jump.xEnd, jump.yEnd)
            }
            println("${i}. From $start $path")
        }
        return getCorrectInput(possibleCaptures)
    }
}

class DummyBot(name: String=""): Player(name) {

    override val name = if(name == "") { NAMES[(Math.random() * NAMES.size).toInt()] } else { name }

    override fun move(possibleMoves: List<Move>, board: Board): Int {
        return (Math.random() * possibleMoves.size).toInt()
    }

    override fun capture(possibleCaptures: List<List<Jump>>, board: Board): Int {
        return (Math.random() * possibleCaptures.size).toInt()
    }

    companion object {
        private val NAMES = arrayOf(
            "Niedorzeczny Bóbr",
            "Zgryźliwa Żmija",
            "Szybki Sebastian",
            "Pijany Olek",
            "Potężna Grochówka",
            "Smutny Marek",
            "Wesoły Romek"
        )
    }

}

class MinMaxBot(name:String, val searchDepth:Int, val estimator: Estimator): Player(name) {
    override val name = if(name == "") { NAMES[(Math.random() * NAMES.size).toInt()] } else { name }

    constructor(estimator: Estimator, searchDepth: Int=3): this("", searchDepth, estimator)

    override fun move(possibleMoves: List<Move>, board: Board): Int {
        val bestMoves = mutableListOf<Int>()
        var bestEval = Int.MIN_VALUE
        val startTime = currentTimeMillis()
        for((i, move) in possibleMoves.withIndex()) {
            var newBoard = board.copy()
            newBoard = CheckersController.executeMove(newBoard, move)
            val eval = minmax(newBoard, searchDepth, this.colour.oppositeColour(), false)
            if(eval > bestEval) {
                bestEval = eval
                bestMoves.clear()
                bestMoves.add(i)
            } else if(eval == bestEval) {
                bestMoves.add(i)
            }
        }
        val endTime = currentTimeMillis()
        println("Computed in ${(endTime - startTime) / 1000}s")
        println("Best estimations for $colour is $bestEval")
        return bestMoves[(Math.random() * bestMoves.size).toInt()]
    }

    override fun capture(possibleCaptures: List<List<Jump>>, board: Board): Int {
        val bestCaptures = mutableListOf<Int>()
        var bestEval = Int.MIN_VALUE
        val startTime = currentTimeMillis()
        for((i, captureList) in possibleCaptures.withIndex()) {
            var newBoard = board.copy()
            newBoard = CheckersController.executeCapture(newBoard, captureList)
            val eval = minmax(newBoard, searchDepth, this.colour.oppositeColour(), false)
            if(eval > bestEval) {
                bestEval = eval
                bestCaptures.clear()
                bestCaptures.add(i)
            } else if(eval == bestEval) {
                bestCaptures.add(i)
            }
        }
        val endTime = currentTimeMillis()
        println("Computed in ${(endTime - startTime) / 1000}s")
        println("Best estimations for $colour is $bestEval")
        return bestCaptures[(Math.random() * bestCaptures.size).toInt()]
    }

    fun minmax(board: Board, depth: Int, currentColour: CheckersColour, maximising: Boolean): Int {
        if(depth == 0) {
            return estimator(board, this.colour)
        }
        val allCaptures = CheckersController.getAllCaptures(board, currentColour)
        if(allCaptures.isNotEmpty()) {
            return minmaxJumps(board, allCaptures, depth, currentColour, maximising)
        }
        val allMoves = CheckersController.getAllMoves(board, currentColour)
        if(allMoves.isNotEmpty()) {
            return minmaxMoves(board, allMoves, depth, currentColour, maximising)
        }
        val currentEstimation = estimator(board, this.colour, true)
        if(maximising) {
            return currentEstimation + searchDepth - depth
        } else {
            return currentEstimation - searchDepth + depth
        }
    }

    fun minmaxMoves(board: Board, moves: List<Move>, depth: Int, currentColour: CheckersColour, maximising: Boolean): Int {
        if(maximising) {
            var currentBest = Int.MIN_VALUE
            for(move in moves) {
                var newBoard = board.copy()
                newBoard = CheckersController.executeMove(newBoard, move)
                val est = minmax(newBoard, depth - 1, currentColour.oppositeColour(), false)
                if(est > currentBest) {
                    currentBest = est
                }
            }
            return currentBest
        } else {
            var currentWorst = Int.MAX_VALUE
            for(move in moves) {
                var newBoard = board.copy()
                newBoard = CheckersController.executeMove(newBoard, move)
                val est = minmax(newBoard, depth -1, currentColour.oppositeColour(), true)
                if(est < currentWorst) {
                    currentWorst = est
                }
            }
            return currentWorst
        }
    }

    fun minmaxJumps(board: Board, possibleJumps:List<List<Jump>>, depth: Int, currentColour: CheckersColour, maximising: Boolean): Int {
        if(maximising) {
            var currentBest = Int.MIN_VALUE
            for(jumps in possibleJumps) {
                var newBoard = board.copy()
                newBoard = CheckersController.executeCapture(newBoard, jumps)
                val est = minmax(newBoard, depth - 1, currentColour.oppositeColour(), false)
                if(est > currentBest) {
                    currentBest = est
                }
            }
            return currentBest
        } else {
            var currentWorst = Int.MAX_VALUE
            for(jumps in possibleJumps) {
                var newBoard = board.copy()
                newBoard = CheckersController.executeCapture(newBoard, jumps)
                val est = minmax(newBoard, depth - 1, currentColour.oppositeColour(), true)
                if(est < currentWorst) {
                    currentWorst = est
                }
            }
            return currentWorst
        }
    }

//    fun minmax2(board: Board, depth: Int, lastColour: CheckersColour, maximising: Boolean): Int {
//        if(CheckersController.hasGameEnded(board, lastColour)) {
//            //println("$lastColour has lost")
////            return estimator(board, lastColour)
//            return if(lastColour == this.colour) {
//                if(maximising) {
//                    Int.MIN_VALUE
//                } else {
//                    Int.MAX_VALUE
//                }
//            } else {
//                if(maximising) {
//                    Int.MAX_VALUE
//                } else {
//                    Int.MIN_VALUE
//                }
//            }
//        }
//        if(depth <= 0) {
//            return estimator(board, this.colour)
//        }
//        val allCaptures = CheckersController.getAllCaptures(board, colour)
//        if(allCaptures.isNotEmpty()) {
//            if(maximising) {
//                var maxEval = Int.MIN_VALUE
//                for(capture in allCaptures) {
//                    var newBoard = board.copy()
//                    newBoard = CheckersController.executeCapture(newBoard, capture)
//                    val eval = minmax2(newBoard, depth - 1, colour.oppositeColour(), false)
//                    maxEval = max(maxEval, eval)
//                }
//                return maxEval
//            } else {
//                var minEval = Int.MAX_VALUE
//                for(capture in allCaptures) {
//                    var newBoard = board.copy()
//                    newBoard = CheckersController.executeCapture(newBoard, capture)
//                    val eval = minmax2(newBoard, depth - 1, colour.oppositeColour(), true)
//                    minEval = min(minEval, eval)
//                }
//                return minEval
//            }
//        }
//        val allMoves = CheckersController.getAllMoves(board, colour)
//        if(maximising) {
//            var maxEval = Int.MIN_VALUE
//            for(move in allMoves) {
//                var newBoard = board.copy()
//                newBoard = CheckersController.executeMove(newBoard, move)
//                val eval = minmax2(newBoard, depth - 1, colour.oppositeColour(), false)
//                maxEval = max(maxEval, eval)
//            }
//            return maxEval
//        } else {
//            var minEval = Int.MAX_VALUE
//            for(move in allMoves) {
//                var newBoard = board.copy()
//                newBoard = CheckersController.executeMove(newBoard, move)
//                val eval = minmax2(newBoard, depth - 1, colour.oppositeColour(), true)
//                minEval = min(minEval, eval)
//            }
//            return minEval
//        }
//    }


    companion object {
        private val NAMES = arrayOf(
            "Niedorzeczny Bóbr",
            "Zgryźliwa Żmija",
            "Szybki Sebastian",
            "Pijany Olek",
            "Potężna Grochówka",
            "Smutny Marek",
            "Wesoły Romek"
        )
    }
}

fun main(args: Array<String>) {
    val h = Human("Witek")
    h.move(arrayListOf(Move(0, 0, 1, 1)), Board())
    h.capture(arrayListOf(arrayListOf(Jump(0, 0, 2, 2, 1, 1), Jump(2, 2, 4, 4, 3, 3))), Board())
}